mlflow/pyfunc/scoring_server/client.py

```diff
@@ -49,31 +50,54 @@ def invoke(self, data):
         Invoke inference on input data. The input data must be pandas dataframe or numpy array or
         a dict of numpy arrays.
         """
-        content_type = scoring_server.CONTENT_TYPE_JSON
-
-        def get_jsonable_input(name, data):
-            if isinstance(data, np.ndarray):
-                return data.tolist()
-            else:
-                raise MlflowException(f"Incompatible input type:{type(data)} for input {name}.")
-
-        if isinstance(data, pd.DataFrame):
-            post_data = {"dataframe_split": data.to_dict(orient="split")}
-        elif isinstance(data, dict):
-            post_data = {"inputs": {k: get_jsonable_input(k, v) for k, v in data}}
-        elif isinstance(data, np.ndarray):
-            post_data = ({"inputs": data.tolist()},)
-        else:
-            post_data = data
-        if not isinstance(post_data, str):
-            post_data = json.dumps(post_data, cls=_CustomJsonEncoder)
         response = requests.post(
             url=self.url_prefix + "/invocations",
-            data=post_data,
-            headers={"Content-Type": content_type},
+            data=dump_input_data(data),
+            headers={"Content-Type": scoring_server.CONTENT_TYPE_JSON},
         )
         if response.status_code != 200:
             raise Exception(
                 f"Invocation failed (error code {response.status_code}, response: {response.text})"
             )
         return PredictionsResponse.from_json(response.text)
+
+
+def wait_until_file_exists(f: Path, timeout: int = 30) -> None:
+    begin_time = time.time()
+    while True:
+        if f.exists():
+            return
+        time.sleep(1)
+        if time.time() - begin_time > timeout:
+            raise MlflowException(f"Waiting for file {f} timeout")
+
+
+class StdinScoringServerClient:
+    def __init__(self, process):
+        self.process = process
+        self.tmpdir = Path(tempfile.mkdtemp())
+        self.output_json = self.tmpdir.joinpath("output.json")
+
+    def invoke(self, data):
+        """
+        Invoke inference on input data. The input data must be pandas dataframe or numpy array or
+        a dict of numpy arrays.
+        """
+        request_id = uuid.uuid4().hex
+        request = {
+            "id": request_id,
+            "data": dump_input_data(data),
+            "output_file": str(self.output_json),
+        }
+        self.process.stdin.write(json.dumps(request) + "\n")
+        self.process.stdin.flush()
+        done_file = self.tmpdir.joinpath(f"{request_id}.DONE")
+        wait_until_file_exists(done_file)
```

Comment:
Similar feedback to https://github.com/mlflow/mlflow/pull/7685#discussion_r1073013550, I don't think we should be creating separate files for each request. Instead, can we use a single file for processing the results?

Reply:
Got it, I'll fix this

Reply:
Thanks Haru! :D